### ISSUE TITLE
VelocityAt predictions are in a different reference frame from velocity (now) suffix.

### ISSUE THREAD

**Issue by Dunbaratu** [KSP-KOS/KOS#1766]
The intention was for ``velocityat()`` to use the same meaning of velocity as the various `:velocity` suffixes.  It does not.

I now see what people are complaining about with it.  I always thought it was just that they weren't taking into account how things move to a new position in the future and the relative velocities change.  No it's not just that.  It's also wrong for time=now predictions.

For example, if you take this code:
```
set mun_vel to mun:velocity:orbit.
set vdraw to vecdraw(mun:position, 500000* mun_vel, red, "mun_vel", 1, true).
```
you get an arrow pointing from the mun, in the mun's direction of orbit around kerbin.

But replace that with the velocityat prediction, but passing in a time of "now", which should give you the same answer, or at least an answer within a small margin of error for prediction, and you don't just get a small difference.  You get a huge difference.
```
set now to time:seconds.
set mun_vel to velocityat(mun, now):orbit.
set vdraw to vecdraw(mun:position, 500000* mun_vel, red, "mun_vel", 1, true).
```
Now you have a velocity vector that can be as much as 60-70 degrees different from what you had last time.

There is definitely something wrong with velocityat(), and whatever it is, it will be a painful fix because it will definitely be breaking for anyone who got used to the old way and corrected for it.  It does not match my intentions, which is that when given a timestamp of "right now", it would be the same (or close to it within an error margin) as the immediate version of the same data.

**Comment by hany** [comment]
While this is a very old issue, I believe I’m running into this problem. For a stable orbit and a fixed point in time (say `time:seconds + eta:apoapsis`), running `velocityat` will return different results depending on the current velocity of the ship. For example, running this while near the periapsis:

```
print eta:apoapsis.
1458.2850
print velocityat(ship, time:seconds + eta:apoapsis):orbit.
V(834.2167, 0.0007, 219.1011)
print velocityat(ship, time:seconds + eta:apoapsis):orbit:mag.
862.5099
```

Then when warping ahead to a position closer to the apoapsis:

```
print eta:apoapsis.
259.5085
print velocityat(ship, time:seconds + eta:apoapsis):orbit.
V(853.3015, 0.0007, 125.6952).
print velocityat(ship, time:seconds + eta:apoapsis):orbit:mag.
862.5099
```

The magnitude of the predicted velocity is the same in each case, but the direction is incorrect and seems to be dependent on the current position of the `ship` in the orbit. Since in both cases, it’s referring to the same point in time, shouldn’t `velocityat` return the same vector in both cases?

The original issue mentions:

> There is definitely something wrong with velocityat(), and whatever it is, it will be a painful fix because it will definitely be breaking for anyone who got used to the old way and corrected for it.

I’m curious how one would go about correcting for this?

**Comment by nuggreat** [comment]
What you are seeing is not related to this issue.

As what you are seeing is indeed correct behavior and not a bug as KSP can and does rotate the unit vectors that define it's coordinate system depending on a number of factors.  As such it is always the recommendation to not directly reference the x,y,z values of a vector and to treat any vectors you get from kOS as if they are only good for the short term as if you save them with out changing reference frames as time passes they will become increasingly incorrect until like a broken clock the circle back around to being correct.

**Comment by hany** [comment]
Thanks for the quick reply. What led me down this path is attempting to write a landing script with predictions within an atmosphere (i.e.: using a simulation with drag and gravity acceleration steps, similar to your scripts). If I ignore the raw values that `velocityat` returns for me as you suggest, and instead use what it provides to predict where I will land, I’m finding that the values I get vary greatly depending on where I currently am in the orbit. If I’m 180 degrees behind from the starting position, I find that that velocity calculations show that I stand still (as if the velocities cancel out). If I run the same simulation when I’m 90 degrees behind, my velocity calculations show that I increase in altitude (since at 90 degrees, the velocity is perpendicular).

Same point in time, different results depending on my current position along the orbit.

**Comment by nuggreat** [comment]
I would need to see code to try to figure that one out but at a guess something about how you are evaluating the results is wrong or something about your setting the initial state of the simulation is incorrect.

**Comment by hany** [comment]
Makes sense. I appreciate any help you can give me. Below is the code I'm currently using to predict. Running the same piece of code at various current positions around the same stable orbit gives different results (it is simply being called by `predict_landing(CdA)`:

```
function predict_landing {
	parameter CdA.
	parameter target_alt is 0.
	parameter init_pos is ship:position.
	parameter init_vel is ship:velocity:surface.
	parameter init_obt is ship:obt.
	parameter ref_body is ship:body.
	parameter sim_step is 0.1.

	local impact_eta is 0.
	local has_impact is false.
	local nd is false.
	local start_time is time:seconds.
	
	// OK, we're not in an aerocapture orbit yet.
	// Let's pretend we altered the orbit to lower the PE to aerocapture height
	if init_obt:periapsis > ref_body:atm:height {
		local deorbit_dv is landDeorbitDeltaV(aerocapture_alts[ref_body:name]).
		set start_time to start_time + eta:apoapsis.
		local nd is node(start_time, 0, 0, deorbit_dv).
		add nd.
		set init_obt to orbitat(ship, start_time + 1).
	}

	local cur_pos is init_pos.
	local cur_vel is 0.
	lock up_vec to cur_pos - ref_body:position.
	lock cur_alt to up_vec:mag - ref_body:radius.

	// Start everything at the atmosphere boundary.
	if cur_alt > ref_body:atm:height {
		local time_to_atm is eta_altitude(ref_body:atm:height - 1, init_obt).
		set cur_pos to positionat(ship, start_time + time_to_atm).
		set cur_vel to velocityat(ship, start_time + time_to_atm):surface.
		set impact_eta to time_to_atm.
	}
	
	// Acceleration.
	lock grav_acc to -(ref_body:mu * up_vec:normalized / up_vec:sqrmagnitude).
	lock drag_acc to get_drag_force(CdA, cur_alt, cur_vel, ref_body).
	lock cur_acc to grav_acc + drag_acc.
	
	lock impact_site to ref_body:geopositionof(cur_pos).

	// Do the simulation.
	// We're assuming constant acceleration during the time slices
	// Ref: https://physics.info/kinematics-calculus/
	until cur_alt <= max(0, impact_site:terrainheight) + target_alt {
		set log_time to log_time + sim_step.
		print "eta:" + round(log_time,1)
			+ " grav:" + round(grav_acc:mag, 2)
			+ " drag:" + round(drag_acc:mag, 2)
			+ " acc:" + round(cur_acc:mag, 2)
			+ " vel:" + round(cur_vel:mag, 2)
			+ " alt:" + round(cur_alt, 2).
		set cur_vel to cur_vel + (cur_acc * sim_step).
		//set cur_pos to cur_pos + (cur_vel * sim_step) + (0.5 * cur_acc * sim_step^2).
		set cur_pos to cur_pos + (cur_vel * sim_step).
		set impact_eta to impact_eta + sim_step.
	}
	set has_impact to true.
	
	if nd {
		remove nd.
	}

	return list(cur_pos, cur_vel, impact_site, impact_eta, has_impact).
}
```

**Comment by nuggreat** [comment]
So the sim it's self looks right, the use of lock is making it slower than it could be but there is nothing wrong with the math assuming the function calls are correct.  Leaving the problem to be found within how you are evaluating the results.  At a guess you are not evaluating the velocity you get from the sim with respect to the position you get from the sim.

**Comment by hany** [comment]
The strange thing is I’m getting strange results within this function call itself. You’ll notice I’m printing the steps each iteration. The altitude is a dead give away. If I run this sim 180 degrees behind the altitude boundary, the altitude stays fixed (as if velocities cancel out). If I run this 90 degrees behind the boundary, the altitude actually goes up and never converges. As you can see, I don’t seem to use any current position variables. The only thing I can think of is `velocityat` is giving me strange vector depending on my position. Even just going through the loop once is showing incorrect results.

**Comment by hany** [comment]
OK, this is completely unrelated to this issue. @nuggreat you hinted at slowness, which is indeed the problem. Running this function takes time to complete, and in the mean time, the ship is continuing to orbit while calculations are run, and `ref_body:position` will keep changing to match (this explains why it was related to my orbit position). I need to set the `ref_body` position once outside the main loop, and use that throughout the loop.

Thanks for your help, and apologies for wasting time on this.

**Comment by cosmozhang1995** [comment]
I don't know if it's related to this bug, but it seems that `orbitat()` returns an orbit with a different orbital velocity with `velocityat()`.

For example I have this code:
```
print "cur speed = " + velocityat(ship, time:seconds + ship:orbit:eta:apoapsis):orbit.
print "cur speed = " + orbitat(ship, time:seconds + ship:orbit:eta:apoapsis):velocity:orbit.
```

And I got:
```
cur speed = V(-1307.36032213081, 11.049552757087, 1703.49533897259)
cur speed = V(2207.72156363567, 6.05266912452256, 489.119613048034)
```

Nether the direction nor the magnitude is the same.

I guess they are in different reference frames. But don't know what are them nor how to compensate it.
